clients/venice-client/src/main/java/com/linkedin/venice/fastclient/DispatchingAvroGenericStoreClient.java

```diff
@@ -378,105 +347,119 @@ protected void streamingBatchGet(
+      Map<String, String> requestHeaders,
+      Function<List<MultiKeyRequestContext.KeyInfo<K>>, byte[]> requestSerializer,
+      MultiKeyStreamingRouteResponseHandler routeResponseHandler) {
+    verifyMetadataInitialized();
     int keyCnt = keys.size();
     if (keyCnt > this.config.getMaxAllowedKeyCntInBatchGetReq()) {
       throw new VeniceKeyCountLimitException(
           getStoreName(),
-          RequestType.MULTI_GET,
+          requestType,
           keyCnt,
           this.config.getMaxAllowedKeyCntInBatchGetReq());
     }
 
     /* Prepare each of the routes needed to query the keys */
     requestContext.instanceHealthMonitor = metadata.getInstanceHealthMonitor();
-    String uriForBatchGetRequest = composeURIForBatchGetRequest(requestContext);
     int currentVersion = requestContext.currentVersion;
     Map<Integer, List<String>> partitionRouteMap = new HashMap<>();
+    Map<String, CompletableFuture<Void>> routeToResponseFutures = new VeniceConcurrentHashMap<>();
+    Set<String> partitionsWithNoRoutes = new HashSet<>();
     for (K key: keys) {
       byte[] keyBytes = keySerializer.serialize(key);
       // For each key determine partition
       int partitionId = metadata.getPartitionId(currentVersion, keyBytes);
       // Find routes for each partition
       List<String> routes = partitionRouteMap.computeIfAbsent(
           partitionId,
-          (partId) -> metadata.getReplicas(
+          (ignored) -> metadata.getReplicas(
               requestContext.requestId,
               currentVersion,
               partitionId,
               1,
-              requestContext.getRoutesForPartitionMapping().getOrDefault(partId, Collections.emptySet())));
+              requestContext.getRoutesForPartitionMapping().getOrDefault(partitionId, Collections.emptySet())));
 
       if (routes.isEmpty()) {
         /* If a partition doesn't have an available route then there is something wrong about or metadata and this is
          * an error */
         requestContext.noAvailableReplica = true;
-        String errorMessage = String.format(
-            "No available route for store: %s, version: %s, partitionId: %s",
-            getStoreName(),
-            currentVersion,
-            partitionId);
-        LOGGER.error(errorMessage);
-        requestContext.setPartialResponseException(new VeniceClientException(errorMessage));
+        partitionsWithNoRoutes.add(String.valueOf(partitionId));
       }
 
       /* Add this key into each route we are going to send request to.
         Current implementation has only one replica/route count , so each key will go via one route.
         For loop is not necessary here but if in the future we send to multiple routes then the code below remains */
       for (String route: routes) {
-        requestContext.addKey(route, key, partitionId);
+        requestContext.addKey(route, key, keyBytes, partitionId);
+        routeToResponseFutures.put(route, new CompletableFuture<>());
```

Comment:
This introduces the 3rd future for each route on top of `routeFuture` (from `post()`), `routeRequestFuture`(from healthcheck) (PS: we need to rename these to be less confusing 🥲, not now probably). 
wondering how the new future helps?

Reply:
Yeah. I see your point about the confusion. We can probably merge the new future and `routeRequestFuture`. Will see if this is feasible. The new future is decoupled from a "route" and helps track missing routes as failures. This allows us to mark the overall future for `multiKeyStreamingRequest` to fail in any erroneous scenarios.

Reply:
I updated the PR to merge the two - `routeRequestFuture` (from healthcheck) and the new future